### PR TITLE
Add Same-URL A/B Testing to Examples

### LINKED ISSUE #2693: transparent A/B on same URL?
https://github.com/cloudflare/cloudflare-docs/blob/2c7376e6aaf67e3778ad17932907f53157bafe7a/products/workers/src/content/examples/ab-testing.md?plain=1#L20

In the A/B example, it references the test as running on `/control/somepath`, and the control URL being `/control/sompath`. Not sure if this was an intentional typo. 

What would happen if I actually need the control to be `/control/somepath`? When it goes to fetch the control response, wouldn't it hit the worker again, and well, infinitely loop?

In my own testing, it seems like I need 3 different paths/URLs to accomplish an A/B test:
- entry URL (entry point for CF worker)
- test URL
- control URL

Am I doing something wrong?

### PR DESCRIPTION
**Pull request by helloimalastair**
Resolves [transparent A/B on same URL?#2693](https://github.com/cloudflare/cloudflare-docs/issues/2693) issue.
This example supports the default A/B testing setup shown [here](https://developers.cloudflare.com/workers/examples/ab-testing), with one key difference. When the `/control` or `/test` paths are accessed directly, cookies are ignored, and the request is forwarded directly to the desired environment on the origin.

Please edit the title and summary as y'all see fit. I mostly just wrote them as a placeholder.

### CODE REVIEW

**Comment by kristianfreeman** [comment]
Thanks! Can you update this to replace the existing A/B test example

**Comment by helloimalastair** [comment]
> Thanks! Can you update this to replace the existing A/B test example

Done!

File: products/workers/src/content/examples/ab-testing.md [hunk 1]
```diff
@@ -1,45 +1,45 @@
...
+  else if (cookie && cookie.includes(`${NAME}=test`)) url.pathname = "/test" + url.pathname;
   else {
     // If there is no cookie, this is a new client. Choose a group and set the cookie.
-    const group = Math.random() < 0.5 ? "test" : "control" // 50/50 split
-    const response = group === "control" ? CONTROL_RESPONSE : TEST_RESPONSE
-    response.headers.append("Set-Cookie", `${NAME}=${group}; path=/`)
-
-    return response
+    const group = Math.random() < 0.5 ? "test" : "control"; // 50/50 split
+    if(group === "control") url.pathname = "/control" + url.pathname;
```
Comment (kristianfreeman) [inline products/workers/src/content/examples/ab-testing.md]: can we do long-form if/else here with `{}`?

File: products/workers/src/content/examples/ab-testing.md [hunk 2]
```diff
@@ -1,45 +1,45 @@
...
   // Determine which group this requester is in.
-  const cookie = request.headers.get("cookie")
-  if (cookie && cookie.includes(`${NAME}=control`)) {
-    return CONTROL_RESPONSE
-  }
-  else if (cookie && cookie.includes(`${NAME}=test`)) {
-    return TEST_RESPONSE
-  }
+  const cookie = req.headers.get("cookie");
+  if (cookie && cookie.includes(`${NAME}=control`)) url.pathname = "/control" + url.pathname;
```
Comment (kristianfreeman) [inline products/workers/src/content/examples/ab-testing.md]: same question as below `if/else` `{}`

File: products/workers/src/content/examples/ab-testing.md [hunk 3]
```diff
@@ -1,45 +1,45 @@
...
-
-  // The Responses below are placeholders. You can set up a custom path for each test (e.g. /control/somepath ).
-  const TEST_RESPONSE = new Response("Test group") // e.g. await fetch("/test/sompath", request)
-  const CONTROL_RESPONSE = new Response("Control group") // e.g. await fetch("/control/sompath", request)
+const NAME = "myExampleWorkersABTest";
 
+async function ABTestingWithPassthrough(req) {
+  const url = new URL(req.url);
+  // Enable Passthrough to allow direct access to control and test routes.
+  if(url.pathname.startsWith("/control") || url.pathname.startsWith("/test")) return fetch(req);
```
Comment (kristianfreeman) [inline products/workers/src/content/examples/ab-testing.md]:
```suggestion
  if (url.pathname.startsWith("/control") || url.pathname.startsWith("/test")) return fetch(req);
```

**Review by kristianfreeman — CHANGES_REQUESTED** [review]

**Review by kristianfreeman — APPROVED** [review]

**Comment by kristianfreeman** [comment]
@jkup can you review this when you have time?

File: products/workers/src/content/examples/ab-testing.md [hunk 4]
```diff
@@ -1,45 +1,50 @@
...
+    return res;
   }
+  return fetch(url);
 }
 
-addEventListener("fetch", event => {
-  event.respondWith(handleRequest(event.request))
-})
-```
+addEventListener('fetch', e => {
```
Comment (WalshyDev) [inline products/workers/src/content/examples/ab-testing.md]: Double quotes, keep it consistent

File: products/workers/src/content/examples/ab-testing.md [hunk 5]
```diff
@@ -1,45 +1,50 @@
...
   if (cookie && cookie.includes(`${NAME}=control`)) {
-    return CONTROL_RESPONSE
-  }
-  else if (cookie && cookie.includes(`${NAME}=test`)) {
-    return TEST_RESPONSE
-  }
-  else {
+    url.pathname = "/control" + url.pathname;
+  } else if {
+    (cookie && cookie.includes(`${NAME}=test`)) url.pathname = "/test" + url.pathname;
```
Comment (WalshyDev) [inline products/workers/src/content/examples/ab-testing.md]: Else if was broken here

File: products/workers/src/content/examples/ab-testing.md [hunk 6]
```diff
@@ -1,45 +1,50 @@
...
 ```js
-function handleRequest(request) {
-  const NAME = "experiment-0"
-
-  // The Responses below are placeholders. You can set up a custom path for each test (e.g. /control/somepath ).
-  const TEST_RESPONSE = new Response("Test group") // e.g. await fetch("/test/sompath", request)
-  const CONTROL_RESPONSE = new Response("Control group") // e.g. await fetch("/control/sompath", request)
+const NAME = "myExampleWorkersABTest";
 
+async function ABTestingWithPassthrough(req) {
```
Comment (WalshyDev) [inline products/workers/src/content/examples/ab-testing.md]: nitpick: function name not camelCase - may be fine for the docs though

**Review by WalshyDev — CHANGES_REQUESTED** [review]

File: products/workers/src/content/examples/ab-testing.md [hunk 7]
```diff
@@ -1,45 +1,50 @@
...
 type: example
-summary: Set up an A/B test by controlling what response is served based on cookies.
+summary: Set up an A/B test by controlling what response is served based on cookies. This version supports passing the request through to test and control on the origin, bypassing random assignment.
 tags:
   - Originless
 pcx-content-type: configuration
 ---
 
-# A/B testing
+# A/B Testing with Same-URL Direct Access
```
Comment (deadlypants1973) [inline products/workers/src/content/examples/ab-testing.md]:
```suggestion
# A/B testing with same-URL direct access
```

**Review by deadlypants1973 — APPROVED** [review]

**Review by jkup — APPROVED** [review]
Looks good!

**Pull request merged**